marian-nmt/marian-dev
# Squashed commit of fixes to graceful shutdown via SIGTERM

### Description

This pull request replaces #660. I believe it fixes all the issues raised by @frankseide with respect to that. 
I created a new PR to get github to trigger a new diff to facilitate final code review.

Currently (master) SIGTERM will already trigger a graceful shutdown/exit (i.e., Marian will save models before exiting upon SIGTERM), but batch generation in the background will not pay attention to it. This can mean considerable delays in shutdown. This PR fixes that by inserting checks for graceful exit requests in the batch generation loop.

In addition, the graceful exit behavior upon SIGTERM can now be disabled (set to the default of just terminating everything without saving or any other cleanup) on the command line with `--sigterm immediate`.
Please add a clear and concise description of the changes.

This PR fixes a bug/adds a new feature/refactorizes the code/does something else.
It is related to issues: #998, #999, ...

List of changes:
- addresses concerns by @frankseide. These are mostly cosmetic, such as removing unneccesary braces; plus having a generic flag for graceful exist rather than tying this to a specific signal flag.
- graceful exit via SIGTERM can now be disabled via a configuration switch (see above).
Added dependencies: none

### How to test

1. Start training, wait for the first update, then send termination signal (i.e., `kill <process id>`). Marian should save the model before exiting.

 2. Start training with `--sigterm immediate`, wait for the first update, then send termination signal. Marian should exit immediately.
 
### Checklist
